OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/optimobile/OptimobileHttpClient.java

```diff
@@ -0,0 +1,118 @@
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+class OptimobileHttpClient {
+
+    private static final Object lock = new Object();
+    private static OptimobileHttpClient instance;
+
+    private final OkHttpClient okHttpClient;
+    private @Nullable String authHeader;
+
+    static OptimobileHttpClient getInstance() {
+        if (instance != null) {
+            return instance;
+        }
+        synchronized (lock) {
+            if (instance == null) {
+                instance = new OptimobileHttpClient();
+            }
+        }
+        return instance;
+    }
+
+    private OptimobileHttpClient() {
+        okHttpClient = this.buildOkHttpClient();
+    }
+
+    private OkHttpClient buildOkHttpClient() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            return new OkHttpClient();
+        }
+
+        //ciphers available on Android 4.4 have intersections with the approved ones in MODERN_TLS, but the intersections are on bad cipher list, so,
+        //perhaps not supported by CloudFlare. On older devices allow all ciphers
+        ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+                .allEnabledCipherSuites()
+                .build();
+
+        return new OkHttpClient.Builder()
+                .connectionSpecs(Collections.singletonList(spec))
+                .build();
+    }
+
+    Response postSync(String url, String data) throws IOException, Optimobile.PartialInitialisationException {
+        RequestBody body = RequestBody.create(data, MediaType.parse("application/json; charset=utf-8"));
+
+        Request.Builder builder = new Request.Builder().post(body);
+        Request request = this.buildRequest(builder, url);
+
+        return this.doSyncRequest(request);
+    }
+
+    Response getSync(String url) throws IOException, Optimobile.PartialInitialisationException {
+        Request.Builder builder = new Request.Builder().get();
+
+        Request request = this.buildRequest(builder, url);
+
+        return this.doSyncRequest(request);
+    }
+
+    void getAsync(String url, Callback callback) throws Optimobile.PartialInitialisationException {
+        Request.Builder builder = new Request.Builder().get();
+        Request request = this.buildRequest(builder, url);
+
+        this.doAsyncRequest(request, callback);
+    }
+
+    private Request buildRequest(Request.Builder builder, String url) throws Optimobile.PartialInitialisationException {
+        if (this.authHeader == null) {
+            OptimoveConfig config = Optimove.getConfig();
+
+            String apiKey = config.getApiKey();
+            String secretKey = config.getSecretKey();
+            if (apiKey == null && secretKey == null) {
+                throw new Optimobile.PartialInitialisationException();
+            }
+
+            this.authHeader = buildBasicAuthHeader(apiKey, secretKey);
```

Comment:
Is this thread safe / does it matter?

Reply:
what could be the issue? if multiple threads set this -- fine. guess could only have problems if each thread had somehow different credentials

Reply:
Yes, likely not a major issue, you can just have redundant work.

Just for background:
https://jenkov.com/tutorials/java-concurrency/java-memory-model.html
https://medium.com/@MrAndroid/java-multithreading-part-1-java-memory-model-fb8e0cfab9d3

Reply:
the work would be really minor though + mb it's better to do double work here than e.g. block request until other thread finishes. im cheekily resolving